### ISSUE TITLE
Caution - Using dropout with model sharding - Revisit random seed generation

### ISSUE THREAD

**Issue by anaprietonem** [ecmwf/anemoi-core#595]
### Is your feature request related to a problem? Please describe.

As raised by @ssmmnn11, current implementation for random seed generation implies that every rank in the model group gets the same random seed. While each rank gets different inputs,  if dropout is activated the elements dropped would the "same" which can lead to unexpected behaviour or patterns in the model between ranks.

While we still want to preserve a way to generate a synced random seed across ranks to generate spatial correlated random numbers for sharding, there could be a secondary random stream that allow to have random seeds different across ranks for layers  like dropout.

**Comment by Ebraheem1** [comment]
Hi,

By diving into the codebase to think about the root cause(s) behind this issue, I think it's also related to #498. I'd like to take this on. I have an ongoing analysis that points to different spots. I'd articulate them on the way so this comment will be edited with additional info whenever I stumble upon more root causes.

### First potential reason [Dataloader workers are seeded with a rank-agnostic base seed]

In file `training/src/anemoi/training/data/multidataset.py`, I see two issues:

- All workers across all processes/GPUs get the same seed. `get_base_seed()` returns the same value regardless of which process or worker is calling it. Thus, stochastic data augmentation will provide identical outcomes across different processes. This is definitely sub-optimal in terms of learning effectiveness.
- Workers within the same GPU are not differentiated from each other. `worker_id` is captured (`self.worker_id = worker_id`) but never mixed into the seed.

In my eyes, the correct seed for worker `w` on global rank `r ` should be a function of `(base_seed, r, w)`, e.g., a hash or arithmetic combination.

**Comment by ssmmnn11** [comment]
Thank you for you comment @Ebraheem1. We have a common base seed to sync model and ensemble groups. The way to tackle this is to use different random streams for operations that require to be synchronised between certain ranks (data loading, model init) and operations where we don't want this kind of synchronisation. 

Regarding the dataloader: all workers shuffle the global sample index, and then each worker picks its chunk.

**Comment by Ebraheem1** [comment]
Thanks for mentioning this dataloader chunk thing @ssmmnn11 . You are definitely right here.

The only thing I can point to now is:

In `training/src/anemoi/training/distributed/strategy.py`, `seed_rnd` derives a single seed from `(model_comm_group_id + 1) * base_seed` and applies it via `pl.seed_everything` on every rank:
```
def seed_rnd(model_comm_group_id: int, global_rank: int) -> None:
    base_seed = get_base_seed()
    initial_seed = base_seed * (model_comm_group_id + 1)
    rnd_seed = pl.seed_everything(initial_seed)
    ...
```
Here, there are three takes:

1. All ranks inside the same `model_comm_group` (tensor-parallel shards) share an identical RNG state, so `nn.Dropout` selects the same indices on every shard. This might lead to undesirable hidden correlations.
2. Different `model_comm_group_ids` only get linearly-spaced seeds (`k * base_seed`), which is a weak mixing function. (Not a big point though).
3. `models/src/anemoi/models/layers/processor.py` prevents dropout entirely if the model is sharded. Is that on purpose?

What I think of to solve that:

- **Rank-local stream** which is unique per `(global_rank, model_comm_group_id, ens_comm_group_id)` tuple. Drives the global Python/NumPy/Torch RNG, which is what `nn.Dropout` reads from. This makes dropout masks independent across shards.

**Comment by ssmmnn11** [comment]
@Ebraheem1, what I meant to say: 

1. we need some correlated random numbers between (certain) ranks
2. we need some un-correlated random numbers between ranks

we have currently 1, but not 2.

To solve that you can:

a) introduce a explicity random stream which is seeded per rank which we would draw random numbers from when we need them to be correlated
b) one stream (e.g. default one - this is what dropout is using) which is independently seeded per rank

the first task would be to introduce a) and make sure this stream is used everywhere, where we want the random numbers to be identical / correlated

**Comment by floriankrb** [comment]
@Ebraheem1  welcome and thanks for diving into this.

Just a small personal note: parts of this comment read like AI-assisted writing to me. That's completely fine and not an issue at all, but I'd personally appreciate a bit of openness about it. If you did use a tool, a quick mention of which one and roughly how (drafting, polishing, bouncing ideas off it, etc.) would be much appreciated. A short summary of your prompts would be a nice bonus but definitely not required.

**Comment by Ebraheem1** [comment]
@ssmmnn11 I think your approach of a then b is cleaner here. I did not think about it this way.

To re-iterate to make sure that I got what you mean:

1. **The first PR** will be somehow reimagining the current `seed_rnd` formula, but onto a dedicated `torch.Generator`. Then change every part in the code that relies on implicit group-wide RNG agreement to draw from that generator explicitly.
2. **The second PR** switch to a rank-local seed that is consumed by `nn.Dropout` and remove the snippet which prevents sharded models from drop-out. 

Did I get that right?

**Comment by Ebraheem1** [comment]
@floriankrb 

you are 100% correct, some of my comments are AI-assisted as it articulates the message I want to convey with details from the codebase in a neat way.

I use Claude usually for many of my work. I start finding issues in repos of interest first. As I am kind of interested to know more about the kind of projects and work done in ECMWF, so I started investigating their open source repos.

How I use it?

I use it primarily to understand the repo structure, this is much faster than going through the codebase dir by dir or module by module, it can help you immediately drawing a diagram of the code module and even diagrams of the modules themselves. This helps understanding the structure. That's the starting point of the discussion about potential files in the codebase that lead to the observed bug mentioned in the issue you link.

Sample Prompt:

```
Analyze this repository with multiple modules: provide a clear overview of each component's role and responsibilities, then neatly visualize their interactions and communication flow using a Mermaid diagram. Also, which ones are responsible for the observed bug in issue # ....
```

Then you take it from there :).
